File: src/main/java/architecture/goldenboughs_lib/client/model/GeoModelExpand.java
package architecture.goldenboughs_lib.client.model;

import architecture.goldenboughs_lib.core.GoldenBoughsLib;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import software.bernie.geckolib.animatable.GeoAnimatable;
import software.bernie.geckolib.cache.GeckoLibCache;
import software.bernie.geckolib.model.GeoModel;

public class GeoModelExpand<T extends GeoAnimatable> extends GeoModel<T> {
	public static final String MODEL_SUFFIX = ".geo.json";
	public static final String TEXTURE_SUFFIX = ".png";
	public static final String ANIMATIONS_SUFFIX = ".animation.json";
	public static final ResourceLocation ITEM_DEFAULT_MODEL = model(GoldenBoughsLib.modRl("item/default")).withSuffix(MODEL_SUFFIX);
	public final ResourceLocation modelPath;
	public final ResourceLocation texturePath;
	public final ResourceLocation animationsPath;

	public GeoModelExpand(ResourceLocation modelPath, ResourceLocation texturePath, ResourceLocation animationsPath) {
		this.modelPath = modelPath.withSuffix(MODEL_SUFFIX);
		this.texturePath = texturePath.withSuffix(TEXTURE_SUFFIX);
		this.animationsPath = animationsPath.withSuffix(ANIMATIONS_SUFFIX);
	}

	public static ResourceLocation model(ResourceLocation path) {
		return path.withPrefix("geo/");
	}

	public static ResourceLocation texturePath(ResourceLocation path) {
		return path.withPrefix("textures/geo/");
	}

	public static ResourceLocation animationsPath(ResourceLocation path) {
		return path.withPrefix("animations/");
	}

	@Override
	public ResourceLocation getModelResource(T animatable) {
		return GeckoLibCache.getBakedModels().get(this.modelPath) == null ? getDefaultModelResource() : this.modelPath;
	}

	@NotNull
	protected ResourceLocation getDefaultModelResource() {
		return ITEM_DEFAULT_MODEL;
	}

	@Override
	public ResourceLocation getTextureResource(T animatable) {
		return texturePath;
	}

	@Override
	public ResourceLocation getAnimationResource(T animatable) {
		return animationsPath;
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/client/model/GeoModelExpand.kt
package architecture.goldenboughs_lib.client.model

import architecture.goldenboughs_lib.api.AllOpe
import architecture.goldenboughs_lib.util.LibUtil
import net.minecraft.resources.ResourceLocation
import software.bernie.geckolib.animatable.GeoAnimatable
import software.bernie.geckolib.cache.GeckoLibCache
import software.bernie.geckolib.model.GeoModel

@AllOpe
class GeoModelExpand<T : GeoAnimatable>(
	modelPath: ResourceLocation,
	texturePath: ResourceLocation,
	animationsPath: ResourceLocation
) : GeoModel<T>() {
	val modelPath: ResourceLocation = modelPath.withSuffix(".geo.json")
	val texturePath: ResourceLocation = texturePath.withSuffix(".png")
	val animationsPath: ResourceLocation = animationsPath.withSuffix(".animation.json")

	protected val defaultModelResource: ResourceLocation = LibUtil.modRl("geo/item/default.geo.json")

	override fun getModelResource(animatable: T?): ResourceLocation =
		if (GeckoLibCache.getBakedModels()[this.modelPath] == null) this.defaultModelResource
		else this.modelPath

	override fun getTextureResource(animatable: T?): ResourceLocation = texturePath

	override fun getAnimationResource(animatable: T): ResourceLocation = animationsPath

	companion object {
		@JvmStatic
		fun modelPath(path: ResourceLocation): ResourceLocation = path.withPrefix("geo/")

		@JvmStatic
		fun texturePath(path: ResourceLocation): ResourceLocation = path.withPrefix("textures/geo/")

		@JvmStatic
		fun animationsPath(path: ResourceLocation): ResourceLocation = path.withPrefix("animations/")
	}
}
